### ISSUE TITLE
Flow does not recognize predicate checks on object properties 

### ISSUE THREAD

**Issue by rmacqueen** [facebook/flow#6097]
Flow supports [annotation of predicates](https://flow.org/en/docs/types/functions/#toc-predicate-functions) to check for possibly undefined values in helper functions. However, this doesn't work if you are checking a property on an object. See this [example](https://flow.org/try/#0PQKgBAAgZgNg9gdzCYAoVBTAHgBzgJwBcxCBPHDMAWVIBVzKBeMAb1TDABMBDQ7gfgBcYAHYBXALYAjDPgDaAXQA0qAL4BudFDEiAxoQCWcEWABusg1FIARXtwAUAZzgSMAeSkArDPuE16FACUwlJwcDAY3CYApLoAFj4A1o6sqACQ+BiEYvgmAIR5zq4e3voAdDx86mBqWjr6RiYGIjDNGADCCbqJ9oGpHKbc+GBYfnQMYMxsHByV3MJyACxKYABMKwDsCuxgGuxpO5Zg9gVYFXZ902AZWTkmZ3NlUAYwhLL23JMAfGCfPwDMgR2qlqqG0ekMxjAum4MBgtDgnSSvX6ZiGIzGASYqNmdgWyzWm22HD210OUGO5nwlhsdnsWEClx2N2yuRG5z4Txeb3wH2+vzAAKBJLUQA) which I include here inline:

```
export type MyType = {
  data?: number[],
};

function verifyData(someObject: MyType): boolean %checks {
	return !!someObject.data; 
}

function inlineCheck() {
  var x: MyType = {
    data: [4, 2, 7]
  };
 	
  if (!!x.data) {
  	return x.data.filter(a => a > 3)
  }
}

function callToCheck() {
  var x: MyType = {
    data: [4, 2, 7]
  };
 	
  if (verifyData(x)) {
  	return x.data.filter(a => a > 3)
  }
}
``` 

Flow does _not_ give me an error for the `inlineCheck()` function, but it _does_ give me an error for the `callToCheck()` function, claiming that the `x.data` field is "possibly undefined". The checks in both cases are identical, the only difference being that in the second case, the check happens in a helper function. Flow should be smart enough to figure out that the `x.data` is defined after I have called `verifyData(a)`.

**Comment by idiostruct** [comment]
In `callToCheck`, the refinement has been invalidated. Here's a workaround: [Try Flow](https://flow.org/try/#0PQKgBAAgZgNg9gdzCYAoVBTAHgBzgJwBcxCBPHDMAWVIBVzKBeMAb1TDABMBDQ7gfgBcYAHYBXALYAjDPgDaAXQA0qAL4BudFDEiAxoQCWcEWAMBnACIYoACgBu3GGIwBKYVLhwYGbiYCkugAWGLoA1masqACQ+BiEYvgmAIRJDk4Y6mBqWjr6RiYGIjCFGADCwWE2LpEcDvhgWMI09BRgzGwcHDx8wnIALEpgAEyDAOwK7GAa7FGTBlBgNilYAHTd3NUdYDFxCSar6ytQBjCEsjbcbQB8YJc3AMwuk6rZqNp6hsZguo4wtHDlEKhKo1MB1BpNOgMNqgrq8bi9AbDMYTDjTbZzBY2cxWWwHeEuTaTHbxRINNbwo4nM74C7XW5gB5PNFqIA).

**Comment by rmacqueen** [comment]
@idiostruct thanks, yes that works. However in the larger context of my application, I need to be able to pass the whole object and have flow know that its doing the check.

**Comment by apsavin** [comment]
@panagosg7 could you help please?

**Comment by panagosg7** [comment]
Unfortunately, adding support for the more general (initial) case is not easy. Here's why:

In @idiostruct's fix, what makes this refinement possible is that at the point of the conditional, Flow knows exactly what the target of the refinement is going to be. In this case this is `x.data`. This allows Flow to replace `x.data` with `x$data`, *assign a fresh type to it*, and enable any potential refinement that `isDef` will bring in.

Note that at the time of this substitution, Flow's inference cannot reliably know what refinement `isDef` performs. Later, when this becomes known, it is simply applied to its argument.

Consider now the case of the of calling `if(verifyData(x)) { ... }`. To us, it is clear that the expression to be refined is `x.data`, but Flow cannot know this at the time it's processing `verifyData(x)`, and so cannot introduce a symbol `x$data` along with a type to be refined.

To sum up, the reason behind this limitation is the mismatch between object refinements, that are performed in a syntactic manner, and function predicates that need to be inferred and so are processed much later.
